Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,446 @@
+            return maxStreams
+        }
+        
+        /// should be called after receiving new http2 settings from the server
+        /// - Parameters:
+        ///   - maxStreams: max streams settings from the server
+        /// - Returns: number of available streams which can be leased
+        mutating func newMaxConcurrentStreams(_ maxStreams: Int) -> Int {
+            guard case let .active(conn, _, usedStreams, lastIdle) = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
+            return max(maxStreams - usedStreams, 0)
+        }
+        
+        mutating func goAwayRecieved() {
+            guard case let .active(conn, maxStreams, usedStreams, _) = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            self.state = .draining(conn, maxStreams: maxStreams, usedStreams: usedStreams)
+        }
+        
+        /// The connection failed to start
+        mutating func failedToConnect() {
+            switch self.state {
+            case .starting:
+                self.state = .backingOff
+            case .backingOff, .active, .draining, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+        
+        mutating func fail() {
+            switch self.state {
+            case .starting, .active, .backingOff, .draining:
+                self.state = .closed
+            case .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func lease(_ count: Int) -> Connection {
+            guard case .active(let conn, let maxStreams, var usedStreams, let lastIdle) = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            usedStreams += count
+            precondition(usedStreams <= maxStreams, "tried to lease a connection which is not available")
+            self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
+            return conn
+        }
+        
+        /// should be called after a request has finished and the stream can be used again for a new request
+        /// - Returns: number of available streams which can be leased
+        mutating func release() -> Int {
+            guard case .active(let conn, let maxStreams, var usedStreams, var lastIdle) = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            usedStreams -= 1
+            assert(usedStreams >= 0)
+            if usedStreams == 0 {
+                lastIdle = .now()
+            }
+            self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
+            return max(maxStreams - usedStreams, 0)
```

Comment:
use switch for the state. If we are draining, we might also release connections.